### PR TITLE
added script to update annots from a Synapse table, datatype specific

### PR DESCRIPTION
**Pull request by kdaily**

### CODE REVIEW

File: update_annotations_from_table.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
+"""This notebook updates all file annotations a project.
```
Comment (kdaily) [inline update_annotations_from_table.py]: not a notebook

File: update_annotations_from_table.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+There are four required file annotations for any assay file uploaded to Synapse.
```
Comment (kdaily) [inline update_annotations_from_table.py]: prerequisites to running

File: update_annotations_from_table.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+This script uses the `dataType` field to get assay-specific files (using a query lookup dictionary `assayToQuery`),
+and then the `UID` to merge existing file annotations and the table metadata (which is obtained through a lookup
+dictionary `assayToMetadataTable`.
+
+"""
+
+import sys
+import pandas as pd
+import numpy
```
Comment (kdaily) [inline update_annotations_from_table.py]: do I need this? it's now there for NAN checking, but maybe in pandas already.

File: update_annotations_from_table.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+logger = logging.getLogger('Annotation Updater')
+logger.setLevel(logging.DEBUG)
+formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
+ch = logging.StreamHandler()
+ch.setLevel(logging.DEBUG)
+ch.setFormatter(formatter)
+logger.addHandler(ch)
+
+from multiprocessing.dummy import Pool
+mp = Pool(10)
```
Comment (kdaily) [inline update_annotations_from_table.py]: parameterized (user-specified)

File: update_annotations_from_table.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+import synapseclient
+
+syn = synapseclient.Synapse(skip_checks=True)
+syn.login(silent=True)
+
+# Taken from https://github.com/larssono/Analysis_helpers/blob/master/synapseHelpers.py
+# Should probably be in synapseclient.synapseutils
+SYNAPSE_PROPERTIES = ['benefactorId', 'nodeType', 'concreteType', 'createdByPrincipalId',
```
Comment (kdaily) [inline update_annotations_from_table.py]: ask Chris or other engineer about the completeness of this list of attributes.

File: update_annotations_from_table.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+                      'createdOn', 'createdByPrincipalId', 'eTag', 'id', 'modifiedOn',
+                      'modifiedByPrincipalId', 'noteType', 'versionLabel', 'versionComment',
+                      'versionNumber', 'parentId', 'description', 's3Token', 'name']
+
+def query2df(queryContent, filterSynapseFields=True, savedSynapseFields = ('id', 'name')):
```
Comment (kdaily) [inline update_annotations_from_table.py]: probably will go to `synapseclient.utils`

File: update_annotations_from_table.py [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+    - `savedSynapseFields`: list of synapse fields not to filter defaults to id, name
+
+    """
+
+    removedProperties = set(SYNAPSE_PROPERTIES) - set(savedSynapseFields)
+
+    try:
+        queryContent = queryContent['results']
+    except TypeError: #If 'results' not present it is a geneartor
+        queryContent = list(queryContent)
```
Comment (kdaily) [inline update_annotations_from_table.py]: really inefficient - don't convert to list, use as generator

File: update_annotations_from_table.py [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+                             defaults to True
+    - `savedSynapseFields`: list of synapse fields not to filter defaults to id, name
+
+    """
+
+    removedProperties = set(SYNAPSE_PROPERTIES) - set(savedSynapseFields)
+
+    try:
+        queryContent = queryContent['results']
+    except TypeError: #If 'results' not present it is a geneartor
```
Comment (kdaily) [inline update_annotations_from_table.py]: spell chkec dockstring

File: update_annotations_from_table.py [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+        queryContent = list(queryContent)
+    for row in queryContent:
+        for key in row.keys():
+            prefix, new_key = key.split('.', 1)
+            item = row.pop(key)
+            row[new_key] = item[0] if type(item) is list else item
```
Comment (kdaily) [inline update_annotations_from_table.py]: look into this statement

File: update_annotations_from_table.py [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+    parser.add_argument("--dry-run", help="Perform the requested command without updating anything in Synapse.",
+                        action="store_true", default=False)
+
+
+    args = parser.parse_args()
+
+    with file(args.config) as f:
+        config = yaml.load(f)
+
+    logger.info(config)
```
Comment (kdaily) [inline update_annotations_from_table.py]: remove debugging

File: update_annotations_from_table.py [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+        # All files
+        logger.info("Getting %s file list" % dataType)
+        fileTbl = query2df(syn.chunkedQuery(dataTypesToQuery[dataType]))
+
+        # Merge metadata and files
+        logger.info("Merging %s" % dataType)
+        merged = doMerge(fileTbl, meta)
+
+        # Transpose the data and convert it to a dictionary, fix individual entries
+        mergedDict = merged.transpose().to_dict()
```
Comment (kdaily) [inline update_annotations_from_table.py]: why transpose here?

Comment (blogsdon) [inline update_annotations_from_table.py]: i don't know!

File: update_annotations_from_table.py [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+
+    """
+
+    logger.info("Updating annotations for %s" % synid)
+
+    o = syn.get(synid, downloadFile=False)
+    a = syn.getAnnotations(o)
+
+    # Make a copy so updates do not affect the original
+    newa = a.copy()
```
Comment (kdaily) [inline update_annotations_from_table.py]: why making a copy? seems that I shouldn't, but I think it was due to something in `updateDict`.

File: update_annotations_from_table.py [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+    metaResults = syn.tableQuery("select * from %s" % metaSchema.id)
+    return metaResults.asDataFrame()
+
+def doMerge(fileTbl, meta):
+    """Merge file table and metadata table together and add index.
+
+    """
+
+    merged = pd.merge(left=fileTbl[['id', 'UID']], right=meta,
+                      how="left", left_on="UID", right_on="UID")
```
Comment (kdaily) [inline update_annotations_from_table.py]: add flexibility for join cols

File: update_annotations_from_table.py (hunk 3 above)
Comment (xindiguo) [inline update_annotations_from_table.py]: Pandas has <code>pandas.isnull()</code> for checking missing values (NaN in numeric arrays, None/NaN in object arrays).

Comment (kdaily) [inline update_annotations_from_table.py]: yes, I do need this.

File: update_annotations_from_table.py (hunk 11 above)
Comment (kdaily) [inline update_annotations_from_table.py]: `to_dict` does column-wise, need a list of dicts per row.

File: update_annotations_from_table.py (hunk 5 above)
Comment (larssono) [inline update_annotations_from_table.py]: SYNAPSE_PROPERTIES  = ['benefactorId', 'nodeType', 'concreteType', 'createdByPrincipalId', 
                       'createdOn', 'createdByPrincipalId', 'eTag', 'id', 'modifiedOn', 
                       'modifiedByPrincipalId', 'noteType', 'versionLabel', 'versionComment', 
                       'versionNumber', 'parentId', 'description', 's3Token', 'name']

Comment (kdaily) [inline update_annotations_from_table.py]: @larssono can this go in the client?

Comment (larssono) [inline update_annotations_from_table.py]: Yepp.  I will add this and the other items that came from SynapseHelpers in the next release.

Comment (kdaily) [inline update_annotations_from_table.py]: Missing `alias`, `projectId`, `fileNameOverride`

**Pull request merged**